File: lambdas/start-chat/startChatContact.js
var AWS = require('aws-sdk');
AWS.config.update({region: process.env.REGION});
var connect = new AWS.Connect();
var SSM = new AWS.SSM();
​
exports.handler = async (event, context, callback) => {
    console.log("Received event: " + JSON.stringify(event));
    var body = JSON.parse(event["body"]);

    var responseFromSSM = await SSM.getParameter({"Name" : "hlc302-connect-instance-id"}).promise();
    var instance_id = responseFromSSM.Parameter.Value;
    console.log(instance_id);

    responseFromSSM = await SSM.getParameter({"Name" : "hlc302-flow-id"}).promise();
    var flow_parameter = responseFromSSM.Parameter.Value;
    console.log(flow_parameter);
​
    startChatContact(body, instance_id, flow_parameter).then((startChatResult) => {
        callback(null, buildSuccessfulResponse(startChatResult));
    }).catch((err) => {
        console.log("caught error " + err);
        callback(null, buildResponseFailed(err));
    });
};

function startChatContact(body, instanceId, contactFlowId) {

      return new Promise(function (resolve, reject) {
        var startChat = {
            "InstanceId": instanceId == "" ? process.env.INSTANCE_ID : instanceId,
            "ContactFlowId": contactFlowId == "" ? process.env.CONTACT_FLOW_ID : contactFlowId,
            "Attributes": {
                "customerName": body["ParticipantDetails"]["DisplayName"]
            },
            "ParticipantDetails": {
                "DisplayName": body["ParticipantDetails"]["DisplayName"]
            }
        };
​
        connect.startChatContact(startChat, function(err, data) {
            if (err) {
                console.log("Error starting the chat.");
                console.log(err, err.stack);
                reject(err);
            } else {
                console.log("Start chat succeeded with the response: " + JSON.stringify(data));
                resolve(data);
            }
        });
​    });
}

function buildSuccessfulResponse(result) {
    const response = {
        statusCode: 200,
        headers: {
            "Access-Control-Allow-Origin": "http://localhost:9000",
            'Content-Type': 'application/json',
            'Access-Control-Allow-Credentials' : true,
            'Access-Control-Allow-Headers':'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token'
        },
        body: JSON.stringify({
            data: { startChatResult: result }
        })
    };
    console.log("RESPONSE" + JSON.stringify(response));
    return response;
}

function buildResponseFailed(err) {
    const response = {
        statusCode: 500,
        headers: {
            "Access-Control-Allow-Origin": "http://localhost:9000",
            'Content-Type': 'application/json',
            'Access-Control-Allow-Credentials' : true,
            'Access-Control-Allow-Headers':'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token'
        },
        body: JSON.stringify({
            data: {
                "Error": err
            }
        })
    };
    return response;
}

File: lambdas/start-chat/startChatContact.py
import boto3
import logging
import json

logger = logging.getLogger()
logger.setLevel(logging.INFO)

SSM = boto3.client('ssm')
connect = boto3.client('connect')

def lambda_handler(event, context):
    logger.info("Received event: " + json.dumps(event))
    body = event['body']
    
    responseFromSSM = SSM.get_parameter(Name="hlc302-connect-instance-id")
    instance_id = responseFromSSM['Parameter']['Value']
    logger.info(instance_id)
    
    responseFromSSM = SSM.get_parameter(Name="hlc302-flow-id")
    flow_parameter = responseFromSSM['Parameter']['Value']
    logger.info(flow_parameter)
    
    json_body = json.loads(body)
    return startChatContact(json_body, instance_id, flow_parameter)
    

def startChatContact(body, instanceId, contactFlowId):
    try: 
        response = connect.start_chat_contact(InstanceId=instanceId, 
            ContactFlowId=contactFlowId, 
            Attributes={"customerName": body['ParticipantDetails']['DisplayName'] },
            ParticipantDetails={"DisplayName": body['ParticipantDetails']['DisplayName']}
        )
        logger.info('Start chat succeeded with the response ' + json.dumps(response))
        return buildSuccessfulResponse(response)
    except BaseException as err: 
        logger.error("Error starting the chat")
        return buildResponseFailed(err)

def buildSuccessfulResponse(result):
     response = {
         "isBase64Encoded": False,
         "statusCode": 200,
         "headers": {
             "Access-Control-Allow-Origin": "*",
            'Content-Type': 'application/json',
            'Access-Control-Allow-Credentials' : "true",
            'Access-Control-Allow-Headers':'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token'
         },
         "body": json.dumps({"data": {"startChatResult": result}})
     }
     logger.info("RESPONSE " + json.dumps(response))
     return response
     
def buildResponseFailed(err):
    response = {
        "isBase64Encoded": False,
        "statusCode": 500,
        "headers": {
            "Access-Control-Allow-Origin": "*",
            'Content-Type': 'application/json',
            'Access-Control-Allow-Credentials' : "true",
            'Access-Control-Allow-Headers':'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token'
        },
        "body": json.dumps({"data": {" Error": err}})
    }
    logger.info("ERROR RESPONSE " + str(err))
    return response